packages/edit-navigation/src/store/actions.js

```diff
@@ -163,57 +92,248 @@ export const saveNavigationPost = serializeProcessing( function* ( post ) {
+		.filter( isSupportedBlock )
+		.filter( ( block ) => ! getRecordIdFromBlock( block ) );
+
+	const tasks = blocksWithoutRecordId.map( () => ( { saveEntityRecord } ) =>
+		saveEntityRecord( 'root', 'menuItem', {
+			title: __( 'Menu item' ),
+			url: '#placeholder',
+			menu_order: 1,
+		} )
+	);
 
-function* batchSave( menuId, menuItemsByClientId, navigationBlock ) {
-	const { nonce, stylesheet } = yield apiFetch( {
-		path: '/__experimental/customizer-nonces/get-save-nonce',
+	const results = await registry
+		.dispatch( coreDataStore )
+		.__experimentalBatch( tasks );
+
+	// Return an updated navigation block with all the IDs in
+	const blockToResult = new Map( zip( blocksWithoutRecordId, results ) );
+	return mapBlocksTree( navigationBlock, ( block ) => {
+		if ( ! blockToResult.has( block ) ) {
+			return block;
+		}
+		return addRecordIdToBlock( block, blockToResult.get( block ).id );
 	} );
-	if ( ! nonce ) {
-		throw new Error();
-	}
+};
 
-	// eslint-disable-next-line no-undef
-	const body = new FormData();
-	body.append( 'wp_customize', 'on' );
-	body.append( 'customize_theme', stylesheet );
-	body.append( 'nonce', nonce );
-	body.append( 'customize_changeset_uuid', uuid() );
-	body.append( 'customize_autosaved', 'on' );
-	body.append( 'customize_changeset_status', 'publish' );
-	body.append( 'action', 'customize_save' );
-	body.append(
-		'customized',
-		computeCustomizedAttribute(
-			navigationBlock.innerBlocks,
-			menuId,
-			menuItemsByClientId
+/**
+ * Updates every menu item where a related block has changed.
+ * Sends a batch request with one PUT /wp/v2/menu-items for every updated menu item.
+ *
+ * @param {Object} navigationBlock A navigation block to find updated menu items in.
+ * @param {number} menuId          Menu ID.
+ * @return {Function} An action creator
+ */
+const batchUpdateMenuItems = ( navigationBlock, menuId ) => async ( {
+	registry,
+	dispatch,
+} ) => {
+	const updatedMenuItems = blocksTreeToAnnotatedList( navigationBlock )
+		// Filter out unsupported blocks
+		.filter( ( { block } ) => isSupportedBlock( block ) )
+		// Transform the blocks into menu items
+		.map( ( { block, parentBlock, childIndex } ) =>
+			blockToMenuItem(
+				block,
+				registry
+					.select( coreDataStore )
+					.getMenuItem( getRecordIdFromBlock( block ) ),
+				getRecordIdFromBlock( parentBlock ),
+				childIndex,
+				menuId
+			)
 		)
+		// Filter out menu items without any edits
+		.filter( ( menuItem ) =>
+			dispatch( applyEdits( menuItem.id, menuItem ) )
+		);
+
+	// Map the edited menu items to batch tasks
+	const tasks = updatedMenuItems.map(
+		( menuItem ) => ( { saveEditedEntityRecord } ) =>
+			saveEditedEntityRecord( 'root', 'menuItem', menuItem.id )
 	);
 
-	return yield apiFetch( {
-		url: ajaxurl || '/wp-admin/admin-ajax.php',
-		method: 'POST',
-		body,
+	await registry.dispatch( coreDataStore ).__experimentalBatch( tasks );
+
+	// Throw on failure. @TODO failures should be thrown in core-data
```

Comment:
Agree with the comment. Also feels like it could be handled as part of the batch, so that we can determine whether the entire batch has an error in one go, rather than iterating through each item.

Fine to handle this in a follow-up.

Reply:
Tracked in https://github.com/WordPress/gutenberg/issues/35048